[SPARK-3468][WebUI] Timeline-View feature

sarutak opened the pull request:
I sometimes trouble-shoot and analyse the cause of long time spending job.

At the time, I find the stages which spends long time or fails, then I find the tasks which spends long time or fails, next I analyse the proportion of each phase in a task.

Another case, I find executors which spends long time for running a task and analyse the details of a task.

In such situation, I think it's helpful to visualize timeline  for each application, job, task and the details of proportion of activity for each task.

I added 3 timeline features to existing Web UI.

[Application Timeline View]
This view shows following things.
- When each executor was added/removed and the reason why it's removed.
- When each job  was started/finished.
- Status of each job.

![screenshot from 2015-04-01 16 49 25](https://cloud.githubusercontent.com/assets/4736016/6936886/e35fd582-d891-11e4-980d-8de13f50e442.png)

[Stage Timeline View]
Similar to Application Timeline View, this view shows following things.
- When each executor was added/removed and the reason why it's removed.
- When each job was started/finished.
- Status of each stage.

![screenshot from 2015-04-01 16 50 59](https://cloud.githubusercontent.com/assets/4736016/6936900/0dca6526-d892-11e4-84a8-efd9037af444.png)

[Task Assignment Timeline View]
This view shows following things.
- When each task started/finished
- How long each task spent and the proportion.
- Status of each task.
- Where each task ran on.

![screenshot from 2015-04-01 16 51 54](https://cloud.githubusercontent.com/assets/4736016/6936910/20fd5acc-d892-11e4-9018-80e463881fc2.png)

All the view above is zoomable by mouse wheel action and scrollable by drag action.

andrewor14 commented:
@sarutak @kayousterhout @vanzin what is the status of this PR? It is one of the largest out there at the moment and it seems to have mostly gone stale. I think others have had reservations about including d3 inside the Spark repository itself. I wonder if it makes it cleaner to implement all these super cool visualizations outside of the Spark project using the statuses exposed by something like #4435.

sarutak commented:
I apologize for my late reply. I had no time until last weekend.
Actually, I'm reconsidering what and how should be visualized and trying implementing.
I could show you a concrete implementation in a few weeks.
Should I close this PR for now and reopen later?

kayousterhout commented:
@sarutak it sounds like the plan is to significantly change the implementation; if that's the case, then yes, closing this PR and opening a new one when the new functionality is ready is the right strategy.

FYI: there's been some effort towards implementing a much more restricted version of this that uses D3: https://issues.apache.org/jira/browse/SPARK-6418.

sarutak commented:
@kayousterhout The basic idea is not changed but I try to use vis.js instead of D3.js because vis.js is easy to build rich timeline view.

This is under development version of the new implementation.

[Timeline view for an application]
![2015-03-23 11 38 50](https://cloud.githubusercontent.com/assets/4736016/6787702/da5b6a06-d151-11e4-89c5-d8d1ba68297f.png)

[Timeline view for a stage]
![2015-03-23 11 40 09](https://cloud.githubusercontent.com/assets/4736016/6787735/f54c25e4-d151-11e4-8a7a-2f6d9b0325be.png)


Actually, I talked with Matei at New York last week and showed the implementation, then got some feedbacks.
One of the feedbacks is that each square which meaning each task should show it's  proportion of duration like you are suggesting at SPARK-6418.

sarutak commented:
This is an image of "Task Assignment Timeline" view. It's almost done.
This view is integrated into existing WebUI, zoomable and scrollable.


![2015-03-25 21 21 25](https://cloud.githubusercontent.com/assets/4736016/6840496/660fe442-d336-11e4-88a0-5f488cb6867f.png)


Now the code of this feature is a little bit messy but I'll cleanup and show the code soon.

sarutak committed:
Added Timeline-View feature for Applications, Jobs and Stages

SparkQA commented:
[Test build #29531 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29531/consoleFull) for   PR 2342 at commit [`082f709`](https://github.com/apache/spark/commit/082f70923f3d6342344b5548bddddd41ee5bbd3c).

sarutak commented:
I've updated the description.

SparkQA commented:
[Test build #29531 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29531/consoleFull) for   PR 2342 at commit [`082f709`](https://github.com/apache/spark/commit/082f70923f3d6342344b5548bddddd41ee5bbd3c).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29531/
Test FAILed.

sarutak committed:
Added vis.min.js, vis.min.css and vis.map to .rat-exclude

SparkQA commented:
[Test build #29534 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29534/consoleFull) for   PR 2342 at commit [`5ce1b21`](https://github.com/apache/spark/commit/5ce1b210fe380c7b0fde8b377a0958dce821fa88).

sarutak committed:
Removed unused setting in timeline-view.css

SparkQA commented:
[Test build #29535 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29535/consoleFull) for   PR 2342 at commit [`a76e569`](https://github.com/apache/spark/commit/a76e569c0c922ba755456278051ec6a550120493).

SparkQA commented:
[Test build #29534 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29534/consoleFull) for   PR 2342 at commit [`5ce1b21`](https://github.com/apache/spark/commit/5ce1b210fe380c7b0fde8b377a0958dce821fa88).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29534/
Test PASSed.

SparkQA commented:
[Test build #29535 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29535/consoleFull) for   PR 2342 at commit [`a76e569`](https://github.com/apache/spark/commit/a76e569c0c922ba755456278051ec6a550120493).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29535/
Test PASSed.

sarutak committed:
Cleanup: Added TimelineViewUtils.scala

SparkQA commented:
[Test build #29583 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29583/consoleFull) for   PR 2342 at commit [`e85e9aa`](https://github.com/apache/spark/commit/e85e9aad17e987a5ce5bd32f8581e20af8d3ea0b).

SparkQA commented:
[Test build #29583 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29583/consoleFull) for   PR 2342 at commit [`e85e9aa`](https://github.com/apache/spark/commit/e85e9aad17e987a5ce5bd32f8581e20af8d3ea0b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29583/
Test PASSed.

sarutak committed:
Merge branch 'master' of git://git.apache.org/spark into timeline-viewer-feature

SparkQA commented:
[Test build #29649 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29649/consoleFull) for   PR 2342 at commit [`d05f2c2`](https://github.com/apache/spark/commit/d05f2c2a5683b0204415c753b401bf7f4f8dccba).

SparkQA commented:
[Test build #29649 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29649/consoleFull) for   PR 2342 at commit [`d05f2c2`](https://github.com/apache/spark/commit/d05f2c2a5683b0204415c753b401bf7f4f8dccba).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29649/
Test PASSed.

sarutak committed:
Cleanuped

SparkQA commented:
[Test build #29666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29666/consoleFull) for   PR 2342 at commit [`9fb522e`](https://github.com/apache/spark/commit/9fb522ed02b5212e87c693943e85ff97634931d8).

sarutak committed:
Cleanuped

SparkQA commented:
[Test build #29667 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29667/consoleFull) for   PR 2342 at commit [`16f82cf`](https://github.com/apache/spark/commit/16f82cfd1fb51961d48c76a6774685e472192b4a).

SparkQA commented:
[Test build #29666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29666/consoleFull) for   PR 2342 at commit [`9fb522e`](https://github.com/apache/spark/commit/9fb522ed02b5212e87c693943e85ff97634931d8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29666/
Test PASSed.

SparkQA commented:
[Test build #29667 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29667/consoleFull) for   PR 2342 at commit [`16f82cf`](https://github.com/apache/spark/commit/16f82cfd1fb51961d48c76a6774685e472192b4a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29667/
Test PASSed.

sarutak committed:
Cleanuped

sarutak committed:
Merge branch 'master' of git://git.apache.org/spark into timeline-viewer-feature

SparkQA commented:
[Test build #29693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29693/consoleFull) for   PR 2342 at commit [`dab7cc1`](https://github.com/apache/spark/commit/dab7cc172bc637c5cb86797c333e040a321685f8).

SparkQA commented:
[Test build #29693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29693/consoleFull) for   PR 2342 at commit [`dab7cc1`](https://github.com/apache/spark/commit/dab7cc172bc637c5cb86797c333e040a321685f8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29693/
Test PASSed.

sarutak committed:
Merge branch 'master' of git://git.apache.org/spark into timeline-viewer-feature

SparkQA commented:
[Test build #29741 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29741/consoleFull) for   PR 2342 at commit [`fcdab7d`](https://github.com/apache/spark/commit/fcdab7d402fe0b5462023816fb69d967b6411b98).

SparkQA commented:
[Test build #29741 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29741/consoleFull) for   PR 2342 at commit [`fcdab7d`](https://github.com/apache/spark/commit/fcdab7d402fe0b5462023816fb69d967b6411b98).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29741/
Test PASSed.

sarutak committed:
Merge branch 'master' of git://git.apache.org/spark into timeline-viewer-feature

SparkQA commented:
[Test build #29864 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29864/consoleFull) for   PR 2342 at commit [`dec85db`](https://github.com/apache/spark/commit/dec85db21c66514fb640ca017aac8ca6c2acf75c).

SparkQA commented:
**[Test build #29864 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29864/consoleFull)**     for PR 2342 at commit [`dec85db`](https://github.com/apache/spark/commit/dec85db21c66514fb640ca017aac8ca6c2acf75c)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29864/
Test FAILed.

sarutak commented:
retest this please.

SparkQA commented:
[Test build #29886 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29886/consoleFull) for   PR 2342 at commit [`dec85db`](https://github.com/apache/spark/commit/dec85db21c66514fb640ca017aac8ca6c2acf75c).

SparkQA commented:
[Test build #29886 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29886/consoleFull) for   PR 2342 at commit [`dec85db`](https://github.com/apache/spark/commit/dec85db21c66514fb640ca017aac8ca6c2acf75c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29886/
Test FAILed.

sarutak commented:
retest this please.

SparkQA commented:
[Test build #29918 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29918/consoleFull) for   PR 2342 at commit [`dec85db`](https://github.com/apache/spark/commit/dec85db21c66514fb640ca017aac8ca6c2acf75c).

SparkQA commented:
[Test build #29918 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29918/consoleFull) for   PR 2342 at commit [`dec85db`](https://github.com/apache/spark/commit/dec85db21c66514fb640ca017aac8ca6c2acf75c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29918/
Test PASSed.

sarutak commented:
@kayousterhout @andrewor14 If you have a chance, could you give me feedbacks? Thanks.

sarutak committed:
Merge branch 'master' of git://git.apache.org/spark into timeline-viewer-feature

sarutak committed:
Merge branch 'timeline-viewer-feature' of github.com:sarutak/spark into timeline-viewer-feature

SparkQA commented:
[Test build #30014 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30014/consoleFull) for   PR 2342 at commit [`68b7540`](https://github.com/apache/spark/commit/68b7540687f191b036301cc707302f0887a6f477).

SparkQA commented:
[Test build #30014 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30014/consoleFull) for   PR 2342 at commit [`68b7540`](https://github.com/apache/spark/commit/68b7540687f191b036301cc707302f0887a6f477).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class SparseMatrix(Matrix):`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30014/
Test PASSed.

sarutak committed:
Merge branch 'master' of git://git.apache.org/spark into timeline-viewer-feature

SparkQA commented:
[Test build #30211 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30211/consoleFull) for   PR 2342 at commit [`19815ae`](https://github.com/apache/spark/commit/19815aeae23d529a3338520deeeb0a6279a066a4).

SparkQA commented:
[Test build #30211 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30211/consoleFull) for   PR 2342 at commit [`19815ae`](https://github.com/apache/spark/commit/19815aeae23d529a3338520deeeb0a6279a066a4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class Assignment(id: Long, cluster: Int)`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30211/
Test PASSed.

syepes commented:
+1

andrewor14 commented:
@sarutak Thanks for the cool graphs! I haven't looked in detail into the implementation yet, but one concern I have is scalability. It's not unheard of in Spark to have thousands of tasks, and in this case we will render a lot of boxes. I believe the number of items in the existing tables already makes the page load slowly, and I would imagine that rendering these boxes will add to that load. Have you tried viewing the UI with many many partitions? It might make sense to display these graphs lazily.

The other thing is that for certain very short tasks, it seems to me that the breakdown of the delays will be difficult to read. We may need to include this breakdown in the hover menu as text.

sarutak committed:
Merge branch 'master' of git://git.apache.org/spark into timeline-viewer-feature

sarutak commented:
@andrewor14 Thanks for taking a look at this PR!
Yes, I wonder what should we do when there are too many tasks. I have some ideas. First is pagination Stage Page and 2nd is enabling this feature by switching on/off and dialog box will appear and warn when too many tasks will render.

```
The other thing is that for certain very short tasks, it seems to me that the breakdown of the delays will be difficult to read. We may need to include this breakdown in the hover menu as text.
```
This timeline view is zoomable so we can see the breakdown of the delays but showing on hover menu is also good idea.

pwendell commented:
@sarutak can you do some sanity testing to see how large this can realistically scale too at the stage level? For instance try with 10 tasks, 100 tasks, 500 tasks, 1000 tasks, etc? I feel for stage graphs we may need to do something lighter weight like the one proposed in #5547. Running this patch with a stage of 10,000 tasks crashed my browser.

This one seems most useful for the top level jobs view - since overall there is less data and we don't need to worry as much about scalability.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,18 +17,137 @@
+                 }'
+           |}
+         """.stripMargin
+      Option(jobEventJsonAsStr)
```
Why is this an `Option`? Can you just return jobEventJsonAsStr?

pwendell replied:
Or just wrap it in a `Seq` if you are expecting it to be a sequence.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,18 +17,137 @@
+      val jobEventJsonAsStr =
+        s"""
+           |{
+           |  'className': 'job application-timeline-object ${classNameByStatus}',
```
it might be nice to factor these out into a utility function:

```
def generateTimelineEventJSON(
  className: String,
  group: String,
  start: Date,
  end: Option[Date],
  title: String)
```

pwendell replied:
Actually on second though, maybe it's more readable as-is.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,18 +17,137 @@
+           |  'start': new Date(${submissionTime}),
+           |  'end': new Date(${completionTime}),
+           |  'content': '<div class="application-timeline-content">Job ${jobId}</div>',
+           |  'title': 'Job ${jobId}\\nStatus: ${status}\\n' +
```
Can this use the Job description (then say the Job ID in parentheses)? Also, can there be a link that will take you back to the job if you click on it? If you need to add anchor tags for each entry in the jobs table... that's fine.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,18 +17,137 @@
   private val startTime: Option[Long] = parent.sc.map(_.startTime)
   private val listener = parent.listener
 
+  private def applicationTimelineView(jobs: Seq[JobUIData], now: Long): Seq[Node] = {
```
Could you do a few things to make this cleaner and easier to follow?
1. Remove the `val listener` field of the `AllJobsPage` class, then instead just access `parent.listener` directly in the `render` function.
2. Refactor the timeline generation code to be more modular. I'd recommend the following functions:
```
# These can be constants at the top level
val EXECUTORS_LEGEND
val JOBS_LEGEND

def makeJobEvent(job: JobUIData): Seq[Node]
// TODO: Consider pushing this into JobProgressListener
case class ExecutorDataUI(startTime: Long, status: String, finishTime: Option[Long], finishReason: Option[String])
def makeExecutorEvent(executor: ExecutorUIData): Seq[Node]
def makeTimeline(jobs: Seq[JobUIData], executors: seq[ExecutorUIData])
```

Then in the `render` function you can access the listener and create all the necessary inputs to those functions.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala:
```diff
@@ -73,8 +73,15 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
   var numFailedStages = 0
 
+  // Executors:
+  val addedExecutors = TreeSet[(Long, ExecutorId)]()
```
Rather than modifying the JobProgressListener, can you use the existing ExecutorsListener? Feel free to modify it if you need to.

pwendell commented:
Hi @sarutak - I did a pretty thorough pass on this and I have some feedback. First some "high level" stuff:

- The overall architecture. I'm +1 on using this timelines library, it seems like it has a pretty nice interface and simplifies a lot. This also largely uses things we already have inside of existing listeners, which is good.
- Using this vs the approach in #5547. I think a good answer here is to use this vis.js library for the jobs page and then use a custom D3-based approach for the stage page, where we need to be careful about scalability to thousands of events (e.g. thousands of tasks). So with that in mind, I'd propose removing the stage functionality for now from this patch and only having the other pages.
- What does it look like if you just display all stages instead of doing jobs-level view? I wonder if that would be better. Then we could give stages in the same job the same color.

There are also some lower level notes inline about code cleanliness, etc.
- It would be good to have a visual line indicating the start of the application.
- It would be nice if you could use <ctrl>+scroll to zoom, so that we could remove the scroll lock. Is this possible with the library?
- It would be nice if I could mouse over a job and then have it highlight the corresponding job on the table below.

Ping @andrewor14 as well case I missed anything.

andrewor14 commented:
Another thing is that if an application has hundreds of executors, the executors bar will be massive. My understanding of the library is that it will stack events if they happen concurrently, and it's likely that the `ExecutorAdded` events will stack. There is not a perfect workaround for this, but maybe it makes sense to display by default only the last N jobs or last M seconds.

sarutak committed:
Improved scalability

SparkQA commented:
[Test build #30679 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30679/consoleFull) for   PR 2342 at commit [`999ccd4`](https://github.com/apache/spark/commit/999ccd4883337cc1ac1a7dfa84c5be4d0306138e).

sarutak commented:
Thanks a lot for the review @pwendell @andrewor14 !
I've noticed the scalability issue you mentioned and I have an idea to improve the scalability.
The latest patch is the version for better scalability.
The task timeline was displayed on my laptop even if the number of tasks is 1000 and 5000.
Are there still any room for the task timeline of this PR?

And I'll reflect the rest of your feedback soon. Thanks!

pwendell commented:
If you want you can keep the stage view for now, but we may decide to remove it before merging this patch.

SparkQA commented:
[Test build #30679 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30679/consoleFull) for   PR 2342 at commit [`999ccd4`](https://github.com/apache/spark/commit/999ccd4883337cc1ac1a7dfa84c5be4d0306138e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30679/
Test PASSed.

sarutak commented:
O.K. I'll keep the stage view only in my branch and remove it from this PR.

sarutak committed:
Merge branch 'timeline-viewer-feature' of github.com:sarutak/spark into timeline-viewer-feature

sarutak committed:
Removed stage timeline

sarutak committed:
Added job/stage description

sarutak committed:
Added link to job/stage box in the timeline in order to move to corresponding row when we click

pwendell commented:
@sarutak will you have time to address the other feedback soon? We are getting close to the merge deadline and I really want to make sure this feature gets in!

sarutak commented:
@pwendell I'll address all of your feedbacks soon. When should I finish by?

pwendell commented:
If you have time to do an iteration in the next day or two, it would be
helpful. We can continue to go back and fourth ideally over the next week
to have it ready by April 31st.

On Wed, Apr 22, 2015 at 6:10 PM, Kousuke Saruta <notifications@github.com>
wrote:

> @pwendell <https://github.com/pwendell> I'll address all of your
> feedbacks soon. When should I finish by?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2342#issuecomment-95379923>.
>

sarutak committed:
Temporary commit

SparkQA commented:
[Test build #30842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30842/consoleFull) for   PR 2342 at commit [`6a91872`](https://github.com/apache/spark/commit/6a91872db9ce531b4008711cb7544f728c4df31e).

SparkQA commented:
[Test build #30842 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30842/consoleFull) for   PR 2342 at commit [`6a91872`](https://github.com/apache/spark/commit/6a91872db9ce531b4008711cb7544f728c4df31e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30842/
Test FAILed.

sarutak committed:
Refactored

SparkQA commented:
[Test build #30859 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30859/consoleFull) for   PR 2342 at commit [`ee7a7f0`](https://github.com/apache/spark/commit/ee7a7f0c9a9618b05b67f50e5698898b343a059d).

sarutak commented:
I almost addressed your feedbacks.

> Using this vs the approach in #5547. I think a good answer here is to use this vis.js library for the jobs page and then use a custom D3-based approach for the stage page, where we need to be careful about scalability to thousands of events (e.g. thousands of tasks). So with that in mind, I'd propose removing the stage functionality for now from this patch and only having the other pages.

O.K. I respect the approach in #5547 for the stage page.

I'm pending following some feedbacks.

> It would be good to have a visual line indicating the start of the application.

Instead of a visual line, in current implementation, we cannot scroll before the time application started.

> It would be nice if you could use +scroll to zoom, so that we could remove the scroll lock. Is this possible with the library?

vis.js doesn't support that feature directly so we need something trick. I'll implement this feature when I get good idea.

> It would be nice if I could mouse over a job and then have it highlight the corresponding job on the table below.

Instead of this, when we click on event box on the timeline, we can move to the corresponding row in the jobs table.

sarutak committed:
Removed unused function

SparkQA commented:
[Test build #30864 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30864/consoleFull) for   PR 2342 at commit [`974a64a`](https://github.com/apache/spark/commit/974a64a5670c8e8a8078b2e81c915e5808424e14).

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala:
```diff
@@ -73,8 +76,13 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
   var numFailedStages = 0
 
+  // Executors:
+  val executors = ListBuffer[ExecutorUIData]()
```
It doesn't make sense for the `JobProgressListener` to keep track of executor data. Can you put these data structures in the in `ExectuorsListneer` instead? It's fine if you need to add a reference to that listener so it's available here.

sarutak committed:
Added scroll limit to Job timeline

SparkQA commented:
[Test build #30867 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30867/consoleFull) for   PR 2342 at commit [`8110acf`](https://github.com/apache/spark/commit/8110acf82426a5da3e7925f9f27cb7042c817746).

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala:
```diff
@@ -17,20 +17,167 @@
+      val stageId = stage.stageId
+      val attemptId = stage.attemptId
+      val name = stage.name
+      val status = {
```
It might be better to add a private[spark] method called `getStatusString` to `StageInfo`.

pwendell replied:
This is very minor, though

SparkQA commented:
[Test build #30867 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30867/consoleFull) for   PR 2342 at commit [`8110acf`](https://github.com/apache/spark/commit/8110acf82426a5da3e7925f9f27cb7042c817746).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30867/
Test FAILed.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,17 +17,170 @@
+      <text x="35px" y="47px">Executor Removed</text>
+    </svg></div>.toString.filter(_ != '\n')
+
+  private def getlastStageDescription(job: JobUIData) = {
```
`getLastStageDescription`

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,17 +17,170 @@
+    }.map { jobUIData =>
+      val jobId = jobUIData.jobId
+      val status = jobUIData.status
+      val jobDescription = getlastStageDescription(jobUIData)
```
This doesn't seem to be working - When I run the following code:

```
sc.makeRDD(1 to 1000, 1000).count()
```

The description in the jobs table says "count at <console>:22" but the timeline event is empty.

pwendell commented on core/src/main/resources/org/apache/spark/ui/static/timeline-view.js:
```diff
@@ -0,0 +1,90 @@
+      return a.value - b.value
+    },
+    editable: false,
+    showCurrentTime: true,
```
I found this to be a little distracting, maybe we should remove it.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -154,6 +305,9 @@ private[ui] class AllJobsPage(parent: JobsTab) extends WebUIPage("") {
         </div>
 
       var content = summary
+      content ++= <h4>Events on Application Timeline</h4> ++
```
Could this say "Event Timeline" and also, can you make it so that you can toggle it (see the approach in #5547)?

pwendell commented:
I did another pass - this is looking pretty good, but I'd like to test it with more workloads. We should actually try to improve our current workload generator to do things like use Spark SQL and have jobs with many concurrent stages. We can move that to a separate issue.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,17 +17,170 @@
+           |  'start': new Date(${submissionTime}),
+           |  'end': new Date(${completionTime}),
+           |  'content': '<div class="application-timeline-content">' +
+           |    '${jobDescription} (Job ${jobId})</div>',
```
Can this link to the job entry in the table? Actually, if possible it would be nice if on hover, we highlighted the row in the table. This should be pretty easy to do with javascript.

pwendell replied:
Oh I think there may be some link actually already? Can you make it so that the pointer icon appears when I mouse over? I.e. it looks clickable.

sarutak replied:
Now pointer icon appears.

pwendell commented on core/src/main/resources/org/apache/spark/ui/static/timeline-view.js:
```diff
@@ -0,0 +1,90 @@
+  setupZoomable("#application-timeline-zoom-lock", applicationTimeline);
+
+  $(".item.range.job.application-timeline-object").each(function() {
+    $(this).click(function() {
```
It would be good to use javascript so there is a pointer when I mouse over the object.

pwendell replied:
Also, it would be nice if, when I mouse over, the border of the table cell gets thicker or is higlighted in some way.

sarutak replied:
I've implemented tooltip using existing tooltip library.

SparkQA commented:
[Test build #30859 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30859/consoleFull) for   PR 2342 at commit [`ee7a7f0`](https://github.com/apache/spark/commit/ee7a7f0c9a9618b05b67f50e5698898b343a059d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30859/
Test FAILed.

SparkQA commented:
[Test build #30864 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30864/consoleFull) for   PR 2342 at commit [`974a64a`](https://github.com/apache/spark/commit/974a64a5670c8e8a8078b2e81c915e5808424e14).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30864/
Test FAILed.

sarutak committed:
Addressed most of Patrics's feedbacks

sarutak commented:
@pwendell Thanks for lots of feedbacks.
Now I leave only toggling function but I'll address it in 24h.

SparkQA commented:
[Test build #30874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30874/consoleFull) for   PR 2342 at commit [`0dc4278`](https://github.com/apache/spark/commit/0dc42781d74a7ce0456aecee8daf7b15500a702e).

SparkQA commented:
[Test build #30874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30874/consoleFull) for   PR 2342 at commit [`0dc4278`](https://github.com/apache/spark/commit/0dc42781d74a7ce0456aecee8daf7b15500a702e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30874/
Test FAILed.

pwendell commented on core/src/main/resources/org/apache/spark/ui/static/timeline-view.js:
```diff
@@ -0,0 +1,117 @@
+
+    $(this).hover(
+      function() {
+        $(getJobId(this)).addClass("corresponding-item-hover");
```
This only seems to work for rows that are white background in the jobs table, not the rows that are gray background (I tried it locally with the UIWorkloadGenerator).

sarutak replied:
Thanks. I fixed this bug.

pwendell commented on core/src/main/resources/org/apache/spark/ui/static/timeline-view.js:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    $(this).click(function() {
+      window.location.href = getJobId(this);
```
After thinking about it more - I think this should link to the actual job page (since we have the highlighting now) rather than the job anchor on this page.

pwendell commented on core/src/main/resources/org/apache/spark/ui/static/timeline-view.js:
```diff
@@ -0,0 +1,117 @@
+  setupZoomable("#application-timeline-zoom-lock", applicationTimeline);
+
+  $(".item.range.job.application-timeline-object").each(function() {
+    var getJobId = function(baseElem) {
```
Can you make it so that when I hover over the element I get a tooltip that has the contents contained in the cell? I noticed the contents are not readable when they job is really short. We are already using a tooltip library in the library, so it should be possible to use the existing one.

sarutak replied:
In current implementation, when you hover the element, title will appear and it has the contents (but not use the tooltip library).

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,17 +17,172 @@
+  private val EXECUTORS_LEGEND =
+    <div class="legend-area"><svg width="200px" height="55px">
+      <rect x="5px" y="5px" width="20px" height="15px"
+            rx="2px" ry="2px" stroke="#97B0F8" fill="#D5DDF6"></rect>
```
Is it possible to create CSS styles that govern this fill? I wanted to play with this coloring but you currently have to go modify the source code everywhere. It would be nice to have styles for each thing here so we could adjust the color in a style sheet.

sarutak replied:
I've tried before but it's difficult because width, those are attributes, not style. It may be a constraint of svg and rect elements.

pwendell replied:
I only really care about the colors (stroke and fill) are you sure it does not work for those?

sarutak replied:
I think stroke and fill can work. I'll address it.

sarutak replied:
Now stroke and fill attributes are move to the CSS file.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,17 +17,172 @@
-  private val startTime: Option[Long] = parent.sc.map(_.startTime)
-  private val listener = parent.listener
+  private val JOBS_LEGEND =
+    <div class="legend-area"><svg width="200px" height="85px">
```
I found there is a lot of extra whitespace here. Can we do the following?

1. set `#application-timeline div.legend-area { margin-top 5px; }`
2. Make these much closer vertically (y = 5, y = 30, y = 55) for the boxes.
3. Just say "Succeeded" instead of "Succeeded Job".
4. Can the bounding box be made smaller? Maybe 150px instead of 200?

If you could do the same for all the legends, that would be great.

pwendell commented on core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala:
```diff
@@ -17,17 +17,172 @@
+      (jobEventJsonAsStrSeq ++ executorEventJsonAsStrSeq).mkString("[", ",", "]")
+
+    <div id="application-timeline">
+      <div class="control-panel">
```
It might be good to give this div a 5 pixel lower margin.

sarutak committed:
Fixed highlight issue

Implemented toggling function

SparkQA commented:
[Test build #30879 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30879/consoleFull) for   PR 2342 at commit [`28714b6`](https://github.com/apache/spark/commit/28714b6b2b49a9d704555ac23d24287d1d7d1baf).

SparkQA commented:
[Test build #30879 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30879/consoleFull) for   PR 2342 at commit [`28714b6`](https://github.com/apache/spark/commit/28714b6b2b49a9d704555ac23d24287d1d7d1baf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30879/
Test PASSed.

sarutak committed:
Merge branch 'master' of https://github.com/apache/spark into timeline-viewer-feature

sarutak committed:
Fixed a little bit bugs

SparkQA commented:
[Test build #30916 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30916/consoleFull) for   PR 2342 at commit [`d3c63c8`](https://github.com/apache/spark/commit/d3c63c84a56041756841dd0706d87c8c808e84d3).

sarutak committed:
Move `stroke` and `fill` attribute of rect elements to css

SparkQA commented:
[Test build #30918 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30918/consoleFull) for   PR 2342 at commit [`b09d0c5`](https://github.com/apache/spark/commit/b09d0c5f76aa1eb2912ef625c4bd0ffa2c729d64).

SparkQA commented:
[Test build #30916 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30916/consoleFull) for   PR 2342 at commit [`d3c63c8`](https://github.com/apache/spark/commit/d3c63c84a56041756841dd0706d87c8c808e84d3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30916/
Test PASSed.

SparkQA commented:
[Test build #30918 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30918/consoleFull) for   PR 2342 at commit [`b09d0c5`](https://github.com/apache/spark/commit/b09d0c5f76aa1eb2912ef625c4bd0ffa2c729d64).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30918/
Test PASSed.

sarutak committed:
Implement tooltip using bootstrap

SparkQA commented:
[Test build #30936 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30936/consoleFull) for   PR 2342 at commit [`ef34a5b`](https://github.com/apache/spark/commit/ef34a5b87f03e3c7f623ed2c4ab53c933bf64fa8).

SparkQA commented:
[Test build #30936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30936/consoleFull) for   PR 2342 at commit [`ef34a5b`](https://github.com/apache/spark/commit/ef34a5b87f03e3c7f623ed2c4ab53c933bf64fa8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30936/
Test PASSed.

sarutak committed:
Merge branch 'master' of https://github.com/apache/spark into timeline-viewer-feature

sarutak committed:
Merge branch 'master' of git://git.apache.org/spark into timeline-viewer-feature

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31229 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31229/consoleFull) for   PR 2342 at commit [`79ac03d`](https://github.com/apache/spark/commit/79ac03db050a082e64c10122d1f477a0130bbac4).

SparkQA commented:
[Test build #31229 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31229/consoleFull) for   PR 2342 at commit [`79ac03d`](https://github.com/apache/spark/commit/79ac03db050a082e64c10122d1f477a0130bbac4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class ExecutorUIData(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31229/
Test PASSed.

sarutak committed:
Fixed conflict

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31529 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31529/consoleFull) for   PR 2342 at commit [`11fe67d`](https://github.com/apache/spark/commit/11fe67dfb043b80137c412e908ba055356d15c17).

pwendell commented:
@sarutak thanks for updating this - I think I can merge it soon. I will create JIRA's for some follow up. Actually it would be nice if you could submit a follow up that includes the stage page as well, since we decided to go with this one rather than the other PR.

sarutak commented:
@pwendell O.K, I'll submit timeline-view feature for the stage page as another PR.

SparkQA commented:
[Test build #31529 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31529/consoleFull) for   PR 2342 at commit [`11fe67d`](https://github.com/apache/spark/commit/11fe67dfb043b80137c412e908ba055356d15c17).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31529/
Test PASSed.
